### ISSUE TITLE
Set on target that haven't the maps in the middle of path

### ISSUE THREAD

**Issue by jorgecuesta** [joeycumines/go-dotnotation#1]
Hi, thanks for the package first. I found and issue when I try to set a nested property in a `map[string]inteface{}` that is empty.

Think on an empty map and you try to set 'nested.prop' it fail because this line: https://github.com/joeycumines/go-dotnotation/blob/master/dotnotation/accessor.go#L29
return an error because `nested` is not there, so I change that with this:
```go
func (p Accessor) Set(target interface{}, key string, value interface{}) error {
  properties := p.parser(key)
  var targetValue interface{} // added this
  for i, property := range properties {
    if i == (len(properties) - 1) {
      // we reached the last property
      return p.setter(target, property, value)
    }
    // attempt to get the next level, so we can set the last property
    var err error
    targetValue, err = p.getter(target, property) // assign to targetValue
    if err != nil {
      nextLevel := make(map[string]interface{}) // create next level as map[string]interface{}
      err = p.setter(target, property, nextLevel) // set to current target the next level
      target = nextLevel // change the target to nextLevel so next iteration will set on it.
      if err != nil {
        return err
      }
    } else {
      target = targetValue // let the code follow the idea of original target
    }
  }
  return errors.New("no properties parsed from key: " + key)
}
```
So basically we are going through an empty map and we will make the middle missing map. If u wish I can make a PR but I'm really new in Go world.

**Comment by joeycumines** [comment]
Hi there @jorgecuesta, I think I get what you're trying to say, basically you don't think it should error if there it traverses past the end of the available data structure, is that correct? It's been a while so I can only guess what my intent was here, but I believe that I decided that it would be unexpected behavior to insert anything between the existing `target` and the given `value`. The design that I went with for that `dotnotation.Accessor` implementation allows arbitrary behavior via the `Getter` `Setter` and `Parser` fields, with the intent to make it straightforward to extend or implement behavior for a specific use case, but utilising this existing implementation. Generic `map[string]interface{}` is somewhat suitable for the `Default*` behavior but not much else, somewhat because it can't gracefully handle traversing missing slices (it would mean that you could get via numeric keys for maps and slice indexes but only set one type).

The real utility in this implementation as it stands is inspecting and manipulating totally arbitrary JSON data in a way that is otherwise extremely painful, but that's actually a only a small percentage of cases, and in most of those it's probably still a bad idea to even attempt set. Generic data structures are nowhere near as easy to interact with as actual types, and explicitly handling generic unmarshaling using `json.RawMessage` to defer processing is often the better choice where it's feasible.

However, if you want this feature, then feel free to have a crack at it, it will at least give you some practice with go. In order for me to consider merging your changes, please add a `Resolver` field to `Accessor` like

```go
// Accessor provides two methods, Get and Set, that can be configured to handle custom data structures via the
// exported fields, Parser, Getter, Setter, and Resolver.
type Accessor struct {
	// Getter returns the property value of a given target, or an error.
	Getter func(target interface{}, property string) (interface{}, error)
	// Setter sets the property value of a given target, to a given value, or returns an error.
	Setter func(target interface{}, property string, value interface{}) error
	// Parser converts a given key into a list of properties to access in order to get or set.
	Parser func(key string) []string
	// Resolver instantiates a new value for a given property of a given target, or returns an error.
	Resolver func(target interface{}, property string) (interface{}, error)
}
```

then implement an unexported resolver method that will (in the same style as the others) pass through to the exported Resolver field if it's not nil otherwise falling back to a DefaultResolver (which you will need to implement). That can then replace `nextLevel := make(map[string]interface{})`. Unit tests too. If you do want to give that a try just send me a PR and I will take a look when I can. I'll give you feedback, at least.

**Comment by jorgecuesta** [comment]
Ey Thanks for your time @joeycumines I understand your point, what I'm trying to do is reduce a full mongodb document with N properties to just a "whitelist" of properties defined in a config file to be inserted in a new structure and stored in another document.

For example: from a doc with `{a: {b: 1}, c: 2, d: false}` and a config file with a whitelist of properties that say `{include: ["a.b", "d"]}` so basically in the other hand I need a result of `{a: {b: 1}, d: false}`

This is my use case, I will give a try with your explanation see if I can get it with a better implementation and let you know. Anyway if you know some kind of solution for what I'm trying to get done I accept any suggestion as well.

Thanks!

**Comment by joeycumines** [comment]
No worries @jorgecuesta, for your use case, since it requires support for arbitrary keys, I'd say just unmarshal it into generic data structures and implement your own whitelisting behavior. Just make sure you use the json.Decoder implementation with the "use numbers" option turned on, if you don't want it to munge the occasional number.

It may be useful for you to take a look at a project I probably created around the same time as this one, go-configger, which is a CLI tool I implemented to merge and convert config files (supports json yaml and env), I did intend it to have whitelist behaviour basically exactly as you describe but it was complex to get both blacklist and whitelist working in a sensible way, so since I only implemented it for a very specific use case in the first place I didn't bother to add whitelisting. It's also not polished and not intended as a lib for others to use (only a CLI tool to use in a unix-y way), but definitely take a look to see what could be involved.

https://github.com/joeycumines/go-configger

**Disclaimer:** it's not awful but it's possible you might pick up some bad habits from that implementation, I didn't have a great grasp on the patterns or standards of go at the time, so the general quality and the fact I didn't know random things like it being the standard to shadow the original value in a type switch make me reluctant to suggest you to look at it as an example
